### PR TITLE
chore: upgrade to ktor v3

### PR DESCRIPTION
**Pull request by anssari1**
# Situation

- Request bodies are not logged. Ktor 2 logging plugin does not work with gzipped content, so we created custom logging plugins. Ktor 3 resolved this issue.
- EJMask does not support masking nested fields

# Action

- Upgraded to Ktor 3
- Removed custom logging plugins
- Updated pagination handling to do away with content decoding
- Removed EJMask

# Testing

- All unit tests fixed
- Examples run

Screenshot 2025-01-26 at 5 29 35 PM

# Results

- Request bodies are logged
- Ktor 3[ performance improvements](https://blog.jetbrains.com/kotlin/2024/10/ktor-3-0/#performance-improvements)
- Removed custom logging plugins

# Notes

### CODE REVIEW

File: core/src/test/kotlin/com/expediagroup/sdk/core/plugin/logging/LogMaskerTest.kt [hunk 1]
```diff
@@ -43,41 +42,70 @@ internal class LogMaskerTest {
...
+
+        @Test
+        fun `given a text with number of 10 digits then do not omit it`() {
+            val number = "0123456789"
+            val actual = mask("{" +
+                "\"number\": \"$number\" \"some_key\":\"some_value\"" +
+                "}", DEFAULT_MASKED_BODY_FIELDS)
+            val expected = "{" +
+                "\"number\": \"$number\" \"some_key\":\"some_value\"" +
+                "}"
```
Comment (jordan-n-schmidt) [inline core/src/test/kotlin/com/expediagroup/sdk/core/plugin/logging/LogMaskerTest.kt]: I think we might want more tests to verify the boundaries of this regex: `\\d{12,19}`

File: core/src/test/kotlin/com/expediagroup/sdk/core/plugin/logging/LogMaskerTest.kt [hunk 2]
```diff
@@ -43,41 +42,70 @@ internal class LogMaskerTest {
...
+            actual = mask("{\"$key\":\"something\"}", DEFAULT_MASKED_BODY_FIELDS)
             expected = "{\"$key\":\"$OMITTED\"}"
             assertThat(actual).isEqualTo(expected)
 
-           actual = mask("{" +
-               " \"$key\":\"something\" " +
-               "}")
+            actual = mask("{" +
+                " \"$key\":\"something\" " +
+                "}", DEFAULT_MASKED_BODY_FIELDS)
```
Comment (jordan-n-schmidt) [inline core/src/test/kotlin/com/expediagroup/sdk/core/plugin/logging/LogMaskerTest.kt]: Can you think of a cleaner way to represent these json strings? Maybe we can use some sort of multiline string. Would that help?

File: core/src/main/kotlin/com/expediagroup/sdk/core/constant/provider/LogMaskingRegexProvider.kt [hunk 3]
```diff
@@ -16,5 +16,13 @@
 package com.expediagroup.sdk.core.constant.provider
 
 internal object LogMaskingRegexProvider {
-    fun getMaskedFieldsRegex(maskedBodyFields: Set<String>) = "(?<=[\"']?(${maskedBodyFields.joinToString("|")})[\"']?:\\s?[\"'])(\\s*[^\"']+\\s*)(?=[\"'])".toRegex()
+    fun getMaskedFieldsRegex(maskedBodyFields: Set<String>, value: String = "[^\\\"]+"): Regex {
+        val fields = maskedBodyFields.joinToString("|")
+        return "\"(${fields})(\\\\*\"\\s*:\\s*\\\\*\")${value}(?:\\\\?\"|)".toRegex()
```
Comment (jordan-n-schmidt) [inline core/src/main/kotlin/com/expediagroup/sdk/core/constant/provider/LogMaskingRegexProvider.kt]: This certainly isn't super intuitive at first glance as most regexes are. Anything you can do to help readability here? This could be an appropriate place for a comment to say what this does in plain English and/or a test with a large set of inputs could help understand this as well.

I'd also recommend a comment for the default value of the `value` argument.

**Review by dtayeh — DISMISSED** [review]

**Comment by dtayeh** [comment]
Logs from local test:
`7:39:15.573 [DefaultDispatcher-worker-8] INFO  com.expediagroup.sdk.core.client.Client - ExpediaSDK: REQUEST: https://test.ean.com/v3/itineraries?token=QldfCGlcUA4FV1FWBwIFVwxDRwdWXBBLDwheZkFdBwVSU15SVwtXABlWCABRSwcHX1wYXFMHVUwEXFRQBFQGUAIPBVIQXlVHVApVC0BrDlBnQ1BWWxYBD3w9YHNxcSceQBZfUFI-UAtZWgMEUAVUAFUVUENDVk0AbApXCwZRBVcLBwEKCRNADUNdORZNEVUOTRNYVQtMCFcPVA1tVhQMRUZQQEMJVgFBRQBfBU0OBhQMUUMJB0ILDVZrBABUCAMXAwNEWwVYXk0AU1wOUQgYbGoWEVpNQ1AGCUZRFFtQSEMXC1EBSAZYBABTRUNVEVQ8TE1IAQpgYCJ6InQ3IhEIXRE7RlUXV0ZeUQddFVEfAFQRWAUFVRFVUFJRDgoMCgpUAlAaUwYfAVoTEVlLQAtWQj5BQVgCDF9dOV8CBAcIVwwPVB5ZBw5SVRENA1lYElFDAUYUEwBdUUoFMWp8RA5FE1FCBgtbTAoJWWtQCVABD2QxIEAXVAQTDQEHAFtSXwRTXEtVQ1tURlhSEAgMU2lTDVkLWwdcBFIfBQ5cWl5bFkxbA1YBABRUURlVAEJWA1dpUAEIV1MDVFAXTRJdVlsDAg0GEUBeXEBFXgxTOkRVFV4FYmF0eiAkfnYrdRBXVAlbWztbVloDVhFnWwoNTWdWBFgKRlhXVlUWVkVSW1sTXRAEAh4LBBNBAVFdFBBdD1dsSg5MQwFdXEoARQhWGRYRW1ZQHwYOUAFfEwJEFUBcWVcTb0IHEUULDF9rCwEMDAoIUA4HB1cbCF0DBxQHAVVWFAECB1MVBQVXAwQJAgtRUABQREEBFhZZDg1nClUJV1JRW1cNUQ4eUFEGXxhRA1UDFV0GV1ZOAVMGUVcAUQYNAVFQFxFKCkdTSkNLOgoGCglVB1ICVAEGFgVQD11LVRFWVD5FWloBWAEIVANLCQAUVw1iVQwRUSIDCUYDJAkBTwBVWwJeUFQLCmkeF1hUBxdvBFZHCxBXVg4DWUAODlsQV109DggJSEQSEQQQB1toEhdAQFdDAAtgenh5KzYsa2JiJG4geT1gbDBpKTV8H1ZbFlYSQx9sVlYABgkxakJWA1dpXQBGBxZcExVYVgwJA0MEAmBBXl0TAQB3UgYRVwRVD1IdAwUEEVZVZk4MXhMHCCRNXg8TAQcjVgceVgEFa1tdD18PUhBDVk0MXA0OEwEkFlN7EwIBRExDARIKVEcHIBUBCzVOWAx6BFxEB1MXBiVGAAdGWx8DEFAGEFJwQwYBYEUIXhRQUBNQIBQGUBREVF1NWkMbElYEHAsgBRwEcEMHehwHIBAKA0caRFEVVgARVnJAUAAmEAtbIwZcRgMGQAMgHQYKF15JUUYGVBBWJxJUCjETXVpGAAdGBSUUVAZIEFQPRQhMHBIEChIBJFJHAHxBBSEREEFCQA5cBEpmXQEOAlUXcF8BBlhLE1tbCwkPUgQPXAEMVw==
METHOD: HttpMethod(value=POST)
COMMON HEADERS
-> Accept: application/json
-> Accept-Charset: UTF-8
-> Accept-Encoding: gzip
-> Authorization: <-- omitted -->
-> Customer-Ip: 127.0.0.1
-> User-Agent: expediagroup-sdk-java-rapid/5.3.2-SNAPSHOT (Java 22; Mac OS X 14.4.1)
-> transaction-id: bc7f10a3-cfb6-415c-aea4-b68fc0cfb354
-> x-sdk-title: rapid-sdk
CONTENT HEADERS
-> Content-Type: application/json
BODY Content-Type: application/json
BODY START
{"email":"john@example.com","phone":{"country_code":"1","number":"5550077","area_code":"487"},"rooms":[{"given_name":"John","family_name":"Smith","smoking":false,"special_request":"Please provide a room with a view of the ocean.","loyalty_id":null,"loyalty":null}],"affiliate_reference_id":"8f712a36-47f1-4176-b8f9-c0d9","hold":false,"payments":[{"type":"customer_card","billing_contact":{"given_name":"John","family_name":"Smith","address":{"country_code":"US","line_1":"555 1st St","line_2":"10th Floor","line_3":"Unit 12","city":"Seattle","state_province_code":"WA","postal_code":"98121"}},"number":"<-- omitted -->","security_code":"<-- omitted -->","expiration_month":"08","expiration_year":"2025","third_party_authentication":null,"enrollment_date":"2018-09-15"}],"affiliate_metadata":"data_point_1:123|data_point2:This is data.","tax_registration_number":"12345678910","traveler_handling_instructions":"Please use the card provided for payment. Avoid cancellation as this is for a corporate traveler. Contact traveler if any issues.","invoicing":null}`

File: core/src/test/kotlin/com/expediagroup/sdk/core/plugin/logging/LogMaskerTest.kt [hunk 4]
```diff
@@ -43,42 +46,68 @@ internal class LogMaskerTest {
...
+                "12345678901",
+                "123456789012",
+                "123456789012345",
+                "1234567890123456789",
+                "12345678901234567890"
+            ).map {
+                Arguments.of(
+                    Named.of(it.length.toString(), it),
+                    Named.of(
+                        if (it.length in (12..19)) "omit it" else "do not omit it",
```
Comment (jordan-n-schmidt) [inline core/src/test/kotlin/com/expediagroup/sdk/core/plugin/logging/LogMaskerTest.kt]: I'm all for parameterizing tests for logic reuse across a high number of cases, but we should try without introducing branches into our tests.

Perhaps you can introduce 2 parameterized tests instead; one for the "omit" case and one for the "not omit" case.  This would leave the branching up to the test framework which as hopefully tested their branches, something which we aren't doing here.

**Review by jordan-n-schmidt — APPROVED** [review]

**Review by dtayeh — APPROVED** [review]
